### PR TITLE
Python 3.13 and dependency cleanup + bump

### PR DESCRIPTION
**Pull request by apchytr**
### **User description**
**Context:** With tensorflow removed we are no longer restricted by the Python version we can support. Additionally, we were pinned to an older version of JAX and still including some unnecessary legacy dependencies. In this PR, we add support for Python 3.13 and consequently support more recent versions of `jax`. This was a non trivial change however and appropriate fixes were included. 

**Description of the Change:** Added the `vmap_method="sequential"` argument to `hermite_renormalized_jax`. Removed the jitting of `BackendJax.conj` as that was causing issues. Added @ziofil fix to `euclidean_to_unitary`. Removed `grpcio`, `nbformat`, `semantic-version`, `thewalrus` as required dependencies. Added `thewalrus` as a `dev` dependency. Restrict our supported `uv` version to `0.5.29`.  Updated `oscillator_eigenstate` to make use of the `b` batched `hermite_renormalized`.

**Benefits:** A more up to date Mr Mustard experience.

___

### **PR Type**
Enhancement

___

### **Description**
- Add Python 3.13 support and update version constraints

- Remove unnecessary dependencies and move thewalrus to dev

- Fix JAX compatibility issues with newer versions

- Update GitHub Actions workflows to use Python 3.13

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["Python 3.10-3.12"] --> B["Python 3.10-3.13"]
  C["Old JAX/Equinox"] --> D["Updated JAX/Equinox"]
  E["Required Dependencies"] --> F["Cleaned Dependencies"]
  G["JAX Compatibility Issues"] --> H["Fixed JAX Methods"]
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Dependencies</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>pyproject.toml</strong><dd><code>Update Python version and dependency constraints</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/640/files#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711">+10/-13</a>&nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Bug fix</strong></td><td><details><summary>4 files</summary><table>
<tr>
  <td><strong>backend_jax.py</strong><dd><code>Remove JIT decorator from conj method</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/640/files#diff-277d6b9d9077473e0fca45c03a82a1503f283442e0b7503da968e5330ec3c242">+0/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>backend_manager.py</strong><dd><code>Replace transpose with swapaxes for unitary gradient</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/640/files#diff-272a316cdf15707ddb7384f13258135f593fc560e21dc19cf17c16cfb5d0aa1c">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>hermite.py</strong><dd><code>Add vmap_method sequential parameter to callbacks</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/640/files#diff-6aac5f7f8566a0219c57e5294e6c100fe128ac085710abde68a9bff8a0005cfd">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>optimizer.py</strong><dd><code>Update jax.tree_util.tree_map to jax.tree.map</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/640/files#diff-e9ada977f913dced580afe9f792577cde50ba64ff06880d57bc30d9e3121a276">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Documentation</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>CHANGELOG.md</strong><dd><code>Add Python 3.13 support changelog entry</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/640/files#diff-2a22f598a15a364508bc9a475e6ec9df31958a753e9401fabaac8df4db5bd853">+3/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Configuration changes</strong></td><td><details><summary>5 files</summary><table>
<tr>
  <td><strong>pre_commit.yml</strong><dd><code>Update Python version to 3.13</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/640/files#diff-6a2e9fb077e57351f4a7e10d03b114e256298babdf06e7e7ae666781a5cf36a1">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>tests_docs.yml</strong><dd><code>Update Python version to 3.13</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/640/files#diff-ab8128ac2570014193e36f1c59681c1089efd9245866bb4d9b5f48a15199b63f">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>tests_jax.yml</strong><dd><code>Update test matrix to use Python 3.13</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/640/files#diff-2931a1ab126ce8e2302bb62d7caabd97f5751f53022953f0843aa419301b2626">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>tests_numpy.yml</strong><dd><code>Update test matrix to use Python 3.13</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/640/files#diff-d01653471e6afa1d9bb01bcfdc93e6cb34fc4d45c6143faa64089be7bd816943">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>upload.yml</strong><dd><code>Update Python version to 3.13</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/640/files#diff-8cf0b5fae548aab9bd49f5020da9c0e35d281984b82b8a5358ffb1c1ae5bec13">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr></tr></tbody></table>

</details>

___

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -76,12 +73,11 @@ doc = [
 ]
 interactive = [
     "ipykernel >=6.21.2, <7.0.0",
-    "ipywidgets >=8.0.4, <9.0.0"
```
Comment (apchytr) [inline pyproject.toml:79]: This is already a required dependency so unnecessary here

File: pyproject.toml [hunk 2]
```diff
@@ -62,6 +58,7 @@ dev = [
     "pytest-cov==3.0.0",
     "pytest==8.0",
     "ruff>=0.12.0",
+    "thewalrus >=0.19.0,<1",
```
Comment (apchytr) [inline pyproject.toml:62]: Removed as a required dependency as we only make use of it in the testing suite

File: pyproject.toml [hunk 3]
```diff
@@ -183,4 +179,5 @@ filterwarnings = "ignore::DeprecationWarning"
 where = ["."]
 
 [tool.uv]
+required-version = "0.5.29"
```
Comment (apchytr) [inline pyproject.toml:183]: @timmysilv is this a good time to bump this up?

Comment (timmysilv) [reply pyproject.toml:183]: I'd like to do it in its own PR (or after we merge repos) so that we can announce it to the team and ask everyone to run `uv self update x.y.z` and have everyone be on the same page. But it should be soon yeah.

**Review by timmysilv — DISMISSED** [review]
this looks good to me!

File: mrmustard/math/jax_vjps/hermite.py [hunk 4]
```diff
@@ -70,6 +71,7 @@ def hermite_renormalized_jax(
             A,
             b,
             c,
+            vmap_method="sequential",
```
Comment (hitarth64) [inline mrmustard/math/jax_vjps/hermite.py]: Is this still necessary if we are getting rid of `parallel=True` in `njit`? Other everything else looks good to me as well.

**Review by hitarth64 — DISMISSED** [review]

File: mrmustard/math/jax_vjps/hermite.py (hunk 4 above)
Comment (apchytr) [reply mrmustard/math/jax_vjps/hermite.py]: Still necessary because the issue is using [map_fn](https://github.com/XanaduAI/MrMustard/blob/d3c134018ea3757ab6e9ca2f48e1d769de2bca65/mrmustard/physics/fock_utils.py#L121). `parallel=True` only really affects `hermite_renormalized_batched` whereas this only applies to `hermite_renormalized`.

Comment (apchytr) [reply mrmustard/math/jax_vjps/hermite.py]: And testing the timings it seems like avoiding `map_fn` is significantly slower so we do want to keep it

Comment (apchytr) [reply mrmustard/math/jax_vjps/hermite.py]: Actually, I updated the method to make use of our batching functionality in hermite_renormalize. For Jax it seems this is consistently faster than using vmap. For numpy it looks like its about this same more or less

Comment (hitarth64) [reply mrmustard/math/jax_vjps/hermite.py]: This answers my question. Thanks!

**Comment by timmysilv** [comment]
it seems you're bumping all our dependencies. is that intended? if not, I find the quickest way to fix it is this:
```bash
git checkout develop uv.lock
uv sync
```
I'd prefer to do those bumps in their own PR if we do want it

**Comment by apchytr** [comment]
> it seems you're bumping all our dependencies. is that intended? if not, I find the quickest way to fix it is this:
> 
> ```shell
> git checkout develop uv.lock
> uv sync
> ```
> 
> I'd prefer to do those bumps in their own PR if we do want it

Yes, this was intentional here. I reverted the uv lock file and bumped up only the packages necessary to support 3.13 i.e. `numba` and `numpy`

**Review by timmysilv — DISMISSED** [review]
love it. very nice to see `tool.uv` forcing the version

File: mrmustard/training/optimizer.py [hunk 5]
```diff
@@ -101,7 +101,7 @@ def make_step(
         loss_value, grads = jax.value_and_grad(cost_fn, argnums=tuple(range(len(by_optimizing))))(
             *by_optimizing,
         )
-        conj_grads = jax.tree_util.tree_map(lambda x: jax.numpy.conj(x), grads)
+        conj_grads = jax.tree.map(lambda x: jax.numpy.conj(x), grads)
```
Comment (heltluke) [inline mrmustard/training/optimizer.py:104]: Is this call backwards compatible? Or do you need some custom logic here to call `tree_util.tree_map` for older versions and `tree.map` for the latest?

**Review by heltluke — APPROVED** [review]

File: mrmustard/training/optimizer.py (hunk 5 above)
Comment (apchytr) [reply mrmustard/training/optimizer.py:104]: `jax.tree_util.tree_map` is just a deprecated alias for `jax.tree.map` so they're equivalent 😃

**Review by timmysilv — APPROVED** [review]

File: mrmustard/training/optimizer.py (hunk 5 above)
Comment (heltluke) [reply mrmustard/training/optimizer.py:104]: OK, and it's been that way since `0.4.33` which is older than the minimum `0.5.1` here, got it.
